File: Dongwoon/Week13/지형이동.java
import java.util.*;
class Solution {
    public static int N, h, cnt = 0;
    public static int[] dr = {-1, 0, 1, 0};
    public static int[] dc = {0, 1, 0 ,-1};
    public static int[][] visited;
    public static PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> Integer.compare(o1[2], o2[2]));
    
    public int solution(int[][] land, int height) {
        int answer = 0;
        N = land[0].length;
        h = height;
        visited = new int[N][N];
        
        pq.offer(new int[] {0, 0, 0});
        
        while(cnt < N * N){
            int[] cur = pq.poll();
            int cr = cur[0];
            int cc = cur[1];
            if(visited[cr][cc] == 1) continue;
            cnt += 1;
            answer += cur[2];
            dfs(land, cr, cc);
        }
        
        return answer;
    }
    
    public static void dfs(int[][] land, int sr, int sc){
        visited[sr][sc] = 1;
        for(int i = 0;i<4;i++){
            int nr = sr + dr[i];
            int nc = sc + dc[i];
            if(nr < 0 || nr >= N || nc < 0 || nc >= N) continue;
            if(visited[nr][nc] == 0){
                int cost = Math.abs(land[nr][nc] - land[sr][sc]);
                if(cost <= h){
                    cnt += 1;
                    dfs(land, nr, nc);
                } else{
                    pq.offer(new int[] {nr, nc, cost});
                }
            }
        }
    }
}

File: UnrequiredOne/Week13/지형 이동.py
import heapq


def is_valid_move(r, c, mr, mc):
    return 0 <= r < mr and 0 <= c < mc

def solution(land, height):
    answer = 0
    row = len(land)
    col = len(land[0])
    visited = [[-1] * col for _ in range(row)]
    direct = [(-1, 0), (0, -1), (1, 0), (0, 1)]
    q = [(0, 0, 0)]
    while q:
        cost, r, c = heapq.heappop(q)
        if visited[r][c] == -1:
            visited[r][c] = cost
        else:
            continue
        for dr, dc in direct:
            nr, nc = r + dr, c + dc
            if is_valid_move(nr, nc, row, col) and visited[nr][nc] == -1:
                new_cost = abs(land[nr][nc] - land[r][c])
                if height < new_cost:
                    heapq.heappush(q, (new_cost, nr, nc))
                else:
                    heapq.heappush(q, (0, nr, nc))

    for i in range(row):
        for j in range(col):
            answer += visited[i][j]

    return answer
